### PR TITLE
CSS-9328 stream connector

### PR DESCRIPTION
**Pull request by kian99**
## Description

Recreates #1251

**NB**: This PR now incorporates the changes from #1269, this is necessary to ensure tests pass. But those changes will cause the Juju CLI to stop working. So this PR should only land after Juju CLI 3.5.4 is released.

This PR addresses an issue where users cannot use commands like `juju debug-log` with JAAS. I have explained the issue in more detail in https://github.com/juju/juju/pull/17606, copy-pasting from there.

>Running `juju debug-log` does the following:
>1. Connects to the Juju controller at `/model/<uuid>/api` creating an API client in the normal fashion, establishing a websocket connection and performing a login request.
>2. It then goes on to establish a new separate connection to the `/model/<uuid>/log` endpoint, eventually establishing a websocket but instead of sending an RPC login request, it uses the apiClient's `ConnectStream` method where an HTTP basic auth header is sent that includes the user's password and includes the macaroons as cookies.
>3. Proceeds to read from the websocket to stream logs.
>
>With JAAS' introduction of login providers and particularly the session login provider, it was realised that there is no mechanism in place to send the session token used with OIDC login in the basic auth header used in the `ConnectStream` method. Because JAAS acts as a proxy layer for all controller and model level calls, JAAS has no way of authenticating requests to `/model/<uuid>/log`.

This PR introduces a new type of proxy called the `streamProxy` that allows for any Juju streaming endpoints to be proxied. Currently only the `/log` endpoint is allowed to be accessed provided that the user has model `read` permission.

Fixes [CSS-9328](https://warthogs.atlassian.net/browse/CSS-9328)

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [x] Covered by unit tests
- [ ] Covered by integration tests

## Test instructions
To test the changes,
- Install the juju client snap from 3.5/edge and bootstrap a controller.
- Add the controller to JIMM and add a model to JIMM and try `juju debug-log`.

## Notes for code reviewers
I recommend reviewing the commits in order.

The CLI test changes were necessitated by [this PR](https://github.com/juju/juju/pull/17667/files#diff-6a09735ead260521301d78800279f9c2c610289ba71c3b362ae95e0dd85a50f4R122-R124) in Juju which enforced more strictness on how the Juju CLI would handle login.

[CSS-9328]: https://warthogs.atlassian.net/browse/CSS-9328?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/jujuapi/api.go [hunk 1]
```diff
@@ -36,9 +36,17 @@ func APIHandler(ctx context.Context, jimm *jimm.JIMM, p Params) http.Handler {
...
 	mux := http.NewServeMux()
-	mux.Handle("/", &jimmhttp.WSHandler{
+	mux.Handle("/{uuid}/api", &jimmhttp.WSHandler{
 		Upgrader: websocketUpgrader,
-		Server:   modelProxyServer{jimm: jimm},
+		Server: &apiProxier{apiServer: apiServer{
+			jimm: jimm,
+		}},
+	})
+	mux.Handle("/{uuid}/", &jimmhttp.WSHandler{
```
Comment (SimoneDutto) [inline internal/jujuapi/api.go]: is this endpoint the correct one for the stream? if not I would suggest to be more specific with the path (e.g. `/{uuid}/stream_path`).
This is to avoid confusion when juju try to use some missing endpoint via jimm and it fallbacks to this

File: internal/jujuapi/streamproxy.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		writeError(err.Error(), errors.CodeModelNotFound)
+		return
+	}
+	if ok, err := checkPermission(ctx, finalPath, user, model.ResourceTag()); err != nil {
+		writeError(err.Error(), errors.CodeUnauthorized)
+		return
+	} else if !ok {
+		writeError(fmt.Sprintf("unauthorized access to endpoint: %s", finalPath), errors.CodeUnauthorized)
+		return
+	}
```
Comment (SimoneDutto) [inline internal/jujuapi/streamproxy.go]: why didn't you put this in the `Authenticate`?

File: internal/jujuapi/streamproxy.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+			String: modelUUID,
+			Valid:  modelUUID != "",
+		},
+	}
+	if err := s.jimm.Database.GetModel(context.Background(), &model); err != nil {
+		zapctx.Error(ctx, "failed to find model", zap.String("uuid", modelUUID), zap.Error(err))
+		return dbmodel.Model{}, fmt.Errorf("failed to find model: %s", err.Error())
+	}
+	return model, nil
+}
```
Comment (SimoneDutto) [inline internal/jujuapi/streamproxy.go]: i feel like this method belongs to JIMM and not in the API layer.
I have created it in JIMM in my branch, if you want it i can refactor yours as well when i create my pr

File: internal/jujuapi/streamproxy.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+// proxyStreams starts a simple proxy for 2 websockets.
+// After starting the proxy we listen for the first error
+// returned and then close both connections before waiting
+// for an error from the second connection.
+func proxyStreams(src, dst base.Stream) {
+	errChan := make(chan error, 2)
+	go func() { errChan <- proxy(src, dst) }()
+	go func() { errChan <- proxy(dst, src) }()
+	<-errChan
```
Comment (SimoneDutto) [inline internal/jujuapi/streamproxy.go]: shouldn't we catch and log these errors?

File: internal/jujuapi/streamproxy.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+func proxy(src base.Stream, dst base.Stream) error {
+	for {
+		var data map[string]any
+		err := src.ReadJSON(&data)
+		if err != nil {
+			return err
+		}
+		err = dst.WriteJSON(data)
+		if err != nil {
```
Comment (SimoneDutto) [inline internal/jujuapi/streamproxy.go]: i think you should use the same approach you used in a previous pr, where you don't return errors for closed connections, but you do for json errors.

File: internal/jujuapi/streamproxy_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	ofganames "github.com/canonical/jimm/v3/internal/openfga/names"
+)
+
+type streamProxySuite struct {
+	websocketSuite
+}
+
+var _ = gc.Suite(&streamProxySuite{})
+
+func (s *streamProxySuite) TestDebugLogs(c *gc.C) {
```
Comment (SimoneDutto) [inline internal/jujuapi/streamproxy_test.go:23]: i suggest to expand the debug log good test with some actual debug logs, if that's possible

File: internal/jujuapi/api.go (hunk 1 above)
Comment (kian99) [reply internal/jujuapi/api.go]: It's the fallback for any model endpoint which we haven't defined, which is valid behaviour for now. If the handler receives an endpoint it doesn't recognise it will return an appropriate error.

File: internal/jujuapi/streamproxy.go (hunk 2 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: Because this is authorization so it deserves to go somewhere else.

File: internal/jujuapi/streamproxy.go (hunk 3 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: I agree, but the current apiServer accepts a concrete JIMM struct rather than a JIMM interface. We need to do a larger refactor to pass in the interface then we can refactor this. So I think it's best left for a follow-up.

File: internal/jujuapi/streamproxy.go (hunk 4 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: I'm inclined to say no because we can't do anything with these errors besides possibly logging them.

File: internal/jujuapi/streamproxy.go (hunk 5 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: I would agree, but I don't think it's really possible for the marshal/unmarshal parts to trigger an error in this scenario. But I'll take another look at this.

File: internal/jujuapi/streamproxy_test.go (hunk 6 above)
Comment (kian99) [reply internal/jujuapi/streamproxy_test.go:23]: For some reason the Juju controller deployed as part of the conn suite doesn't return debug logs. Maybe because there is nothing in the model.

File: internal/jujuapi/api.go (hunk 1 above)
Comment (SimoneDutto) [reply internal/jujuapi/api.go]: i would rather prefer the mux to give a 404 then the wshandler to give back some "cannot upgrade" error.
Don't you think it's best to have the full path here also for clarity and when in the future we need to add new endpoints under this model handler?

File: internal/jujuapi/streamproxy.go (hunk 4 above)
Comment (SimoneDutto) [reply internal/jujuapi/streamproxy.go]: i do think logging the error in this case is valid.
Just imagine the stream proxy not working without any error logs.

File: internal/jujuapi/streamproxy_test.go (hunk 6 above)
Comment (SimoneDutto) [reply internal/jujuapi/streamproxy_test.go:23]: i think that if you issue some command like "deploy application" it should log something.

File: internal/jujuapi/api.go (hunk 1 above)
Comment (kian99) [reply internal/jujuapi/api.go]: Yeah that's fair, I'll switch to being explicit in the handler

File: internal/jujuclient/dial.go [hunk 7]
```diff
@@ -331,7 +340,22 @@ func (c *Connection) Context() context.Context {
...
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	requestHeader := jujuhttp.BasicAuthHeader(names.NewUserTag(user).String(), pass)
+
+	modelTag, ok := c.ModelTag()
+	if !ok {
+		return nil, errors.E(op, "no model found")
+	}
+	conn, err := rpc.Dial(c.ctx, c.ctl, modelTag, path, requestHeader)
```
Comment (ale8k) [inline internal/jujuclient/dial.go:355]: I think a rpc.DialWithBasicAuth would be a nice wrapper instead, and another one called something like rpc.DialNoAuth (or better naming plz aha), then you can put request header into it.

Also just move line 348 above this if you don't choose to do this change , because we compute the requestheader then get the model tag, but we waste the request header.

File: internal/jujuapi/streamproxy.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	controllerStream, err := api.ConnectStream(finalPath, nil)
+	if err != nil {
+		zapctx.Error(ctx, "failed to connect stream", zap.Error(err))
+		writeError(fmt.Sprintf("failed to connect stream: %s", err.Error()), errors.CodeConnectionFailed)
+		return
+	}
+	proxyStreams(clientConn, controllerStream)
+}
+
+func (s streamProxier) getModel(ctx context.Context, modelUUID string) (dbmodel.Model, error) {
```
Comment (ale8k) [inline internal/jujuapi/streamproxy.go]: This doesn't feel right to be done here, we're going to db layer in jujuapi?

File: internal/jujuapi/api_test.go [hunk 9]
```diff
@@ -32,7 +32,7 @@ func (s *apiSuite) SetUpTest(c *gc.C) {
 
 	mux := http.NewServeMux()
 	mux.Handle("/api", jujuapi.APIHandler(ctx, s.JIMM, s.Params))
-	mux.Handle("/model/", jujuapi.ModelHandler(ctx, s.JIMM, s.Params))
+	mux.Handle("/model/", http.StripPrefix("/model", jujuapi.ModelHandler(ctx, s.JIMM, s.Params)))
```
Comment (ale8k) [inline internal/jujuapi/api_test.go:35]: Why strip prefix?

File: internal/jujuapi/streamproxy.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+			Message: msg,
+			Code:    string(code),
+		}
+		err := clientConn.WriteJSON(errResult)
+		if err != nil {
+			zapctx.Error(ctx, "failed to write error message to client", zap.Error(err), zap.Any("client message", errResult))
+		}
+	}
+	user, err := s.jimm.UserLogin(ctx, auth.SessionIdentityFromContext(ctx))
+	if err != nil {
```
Comment (ale8k) [inline internal/jujuapi/streamproxy.go:62]: Nit: you don't tend to leave any space between lines, i.e.,:
```go
f, err := something()
if err != nil {
}
l, err := something2()
if err != nil {
}
g, err := something3()
if err != nil {
}
```
and it can be kinda difficult to read when blocked together like this.

Where as:
```go
f, err := something()
if err != nil {
}

l, err := something2()
if err != nil {
}

g, err := something3()
if err != nil {
}
```

Reads a bit easier.

**Review by ale8k — DISMISSED** [review]
lgtm, some minor comments

File: internal/jujuapi/streamproxy_test.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	conn := s.open(c, &api.Info{ModelTag: s.Model.ResourceTag()}, "bob")
+	defer conn.Close()
+	_, err := common.StreamDebugLog(context.TODO(), conn, common.DebugLogParams{})
+	c.Assert(err, gc.IsNil)
```
Comment (ale8k) [inline internal/jujuapi/streamproxy_test.go:29]: I don't think I fully understand this test, does no error really mean its working?

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (kian99) [reply internal/jujuapi/api_test.go:35]: It's because of the change in [jujuapi/api.go](https://github.com/canonical/jimm/pull/1344/files#diff-fe36fdbf3a8c97b34cc281bca11eb3bba8ff8fe8bfc80b84e06abf48c5870391R39), before we just matched `/` onto the api proxy. But now we have some different endpoints, so we just define the relative paths in there and the at the top level here (and in service.go where we also do this), we strip the prefix before passing on the rest of the path. This [SO post](https://stackoverflow.com/questions/27945310/why-do-i-need-to-use-http-stripprefix-to-access-my-static-files) also probably explains it better than I did.

File: internal/jujuapi/streamproxy.go (hunk 10 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go:62]: Will change.

File: internal/jujuapi/streamproxy.go (hunk 8 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: I agree, Simone mentioned the same thing, pasting my comment here
>I agree, but the current apiServer accepts a concrete JIMM struct rather than a JIMM interface. We need to do a larger refactor to pass in the interface then we can refactor this. So I think it's best left for a follow-up.

File: internal/jujuapi/streamproxy.go (hunk 4 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: Adding logging for them.

File: internal/jujuapi/streamproxy_test.go (hunk 11 above)
Comment (kian99) [reply internal/jujuapi/streamproxy_test.go:29]: It actually does funnily enough, because it tests that we can go all the way from the client through to the Juju controller without any auth errors, etc. It would be better to also test there are some debug logs, but I was always getting an empty list of debug logs (probably because there is no application deployed but I'll take another look at that).

File: internal/jujuclient/dial.go (hunk 7 above)
Comment (kian99) [reply internal/jujuclient/dial.go:355]: I like the idea for `rpc.DialWithBasicAuth` but I think what we could do is refactor the parameter list into an object, then you can manipulate that object before passing it to dial potentially. 

I'd like to leave this for a follow-up, I'll implement the second suggestion.

Comment (ale8k) [reply internal/jujuclient/dial.go:355]: Not a fan of a params object in this scenario, we have two different behaviour dials. Lets just have two dial? I would think wrapping a struct up in params is really more for large amounts of params.

File: internal/jujuapi/streamproxy.go (hunk 8 above)
Comment (ale8k) [reply internal/jujuapi/streamproxy.go]: You can create the method on jimm and call it on concrete for now. Beats doing this here honestly.

Comment (kian99) [reply internal/jujuapi/streamproxy.go]: Will do

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (ale8k) [reply internal/jujuapi/api_test.go:35]: Doesn't stripprefix just remove r.path.URL prefixes? Your handler would work all the same? Are we using r.path.URL in the handler anywhere?

File: internal/jujuapi/streamproxy.go (hunk 10 above)
Comment (ale8k) [reply internal/jujuapi/streamproxy.go:62]: Thanks, I don't mean to be picky it just makes things a bit difficult if we don't have a bit of cushion between things aha.

File: internal/jujuapi/streamproxy_test.go (hunk 11 above)
Comment (ale8k) [reply internal/jujuapi/streamproxy_test.go:29]: Yeah let's add an app and try get some logs somehow... If it's super difficult, leave a comment explaining why we aren't testing any logs coming through please.

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (kian99) [reply internal/jujuapi/api_test.go:35]: I think if we don't use stripprefix then we need to put /model/ in the handler, this just strips the parts that are common.

File: internal/jujuapi/streamproxy.go (hunk 5 above)
Comment (kian99) [reply internal/jujuapi/streamproxy.go]: Moved the stream proxy to the rpc package where it feels more appropriate and now I can reuse that work.

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (ale8k) [reply internal/jujuapi/api_test.go:35]: Can you try without please?

File: internal/jujuapi/api.go (hunk 1 above)
Comment (alesstimec) [inline internal/jujuapi/api.go]: there are other endpoints served by the juju contoller at /{uuid}/...

File: internal/jujuclient/dial.go (hunk 7 above)
Comment (kian99) [reply internal/jujuclient/dial.go:355]: Will leave this up to a follow-up PR. But moved the code around so that the request header is not created unnecessarily.

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (kian99) [reply internal/jujuapi/api_test.go:35]: It would work but it would mean adding /model/ to the routes in `jujuapi.ModelHandler`, I did it like this intentionally.

File: internal/jujuapi/streamproxy_test.go (hunk 11 above)
Comment (kian99) [reply internal/jujuapi/streamproxy_test.go:29]: Deploying an app seems tricky, I took a brief look at the [Deploy Client](https://github.com/juju/juju/blob/3.5/api/client/application/client.go#L126) and I don't want to deploy an actual charm from charmhub in these tests. So I've opted to leave a comment.

File: internal/jujuapi/api.go (hunk 1 above)
Comment (kian99) [reply internal/jujuapi/api.go]: True, I've changed this to be explicit about the endpoint.

File: internal/jimm/model.go [hunk 12]
```diff
@@ -686,6 +687,20 @@ func (j *JIMM) AddModel(ctx context.Context, user *openfga.User, args *ModelCrea
 	return mi, nil
 }
 
+func (j *JIMM) GetModel(ctx context.Context, uuid string) (dbmodel.Model, error) {
```
Comment (alesstimec) [inline internal/jimm/model.go:690]: godoc, please

File: internal/jujuapi/streamproxy_test.go (hunk 6 above)
Comment (kian99) [reply internal/jujuapi/streamproxy_test.go:23]: Will leave this for the follow up as discussed.

File: internal/auth/oauth2.go [hunk 13]
```diff
@@ -334,7 +334,7 @@ func (as *AuthenticationService) MintSessionToken(email string) (string, error)
 func (as *AuthenticationService) VerifySessionToken(token string) (_ jwt.Token, err error) {
 	const op = errors.Op("auth.AuthenticationService.VerifySessionToken")
 	errorFn := func(message string) error {
-		return errors.E(op, message, errors.CodeUnauthorized)
+		return errors.E(op, message, errors.CodeSessionTokenInvalid)
```
Comment (alesstimec) [inline internal/auth/oauth2.go:340]: afaik this is also an important change, right? because it will tell juju cli that the token is invalid and differentiate it from the "regular" unauthorized error coming from the juju controller.. perhaps it deserves a godoc, so that somebody doesn't just change it in the future..

File: internal/cmdtest/jimmsuite.go [hunk 14]
```diff
@@ -150,6 +154,17 @@ func (s *JimmCmdSuite) AddAdminUser(c *gc.C, email string) {
 	c.Assert(err, gc.IsNil)
 }
 
+// SetupCLIAccess should be run at the start of all CLI tests that want to communicate with JIMM.
+// It will ensure a user is created in JIMM's in-memory accounts store that matches the username passed in.
+// This is necessary as Juju's CLI library validates the logged-in user matches the expected user from the accounts store.
+// This function will return a login provider that can be used to handle authentication.
+func (s *JimmCmdSuite) SetupCLIAccess(c *gc.C, username string) api.LoginProvider {
```
Comment (alesstimec) [inline internal/cmdtest/jimmsuite.go:161]: this test change could possibly be done in a separate PR, because it just makes the current PR harder to read and does not seem really related to the main topic of the PR.

File: internal/jimm/model.go (hunk 12 above)
Comment (alesstimec) [reply internal/jimm/model.go:690]: i suppose authorization check for this call is made at the jujuapi level?

File: internal/jimmhttp/websocket.go [hunk 15]
```diff
@@ -65,74 +68,16 @@ func (h *WSHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-	data := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
-	if err := conn.WriteControl(websocket.CloseMessage, data, time.Time{}); err != nil {
-		zapctx.Error(ctx, "cannot write close message", zap.Error(err))
-	}
-}
-
 func writeInternalServerErrorClosure(ctx context.Context, conn *websocket.Conn, err any) {
 	data := websocket.FormatCloseMessage(websocket.CloseInternalServerErr, fmt.Sprintf("%v", err))
 	if err := conn.WriteControl(websocket.CloseMessage, data, time.Time{}); err != nil {
 		zapctx.Error(ctx, "cannot write close message", zap.Error(err))
```
Comment (alesstimec) [inline internal/jimmhttp/websocket.go:77]: hmm.. error or a warning, i don't know..

File: internal/jujuapi/streamproxy.go [hunk 16]
```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 Canonical.
+package jujuapi
```
Comment (alesstimec) [inline internal/jujuapi/streamproxy.go:3]: newline between copyright and package, otherwise we'll get copyright as package godoc..

File: internal/jujuapi/streamproxy.go [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/auth"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimmhttp"
+	"github.com/canonical/jimm/v3/internal/openfga"
+	jimmRPC "github.com/canonical/jimm/v3/internal/rpc"
+)
+
+// A streamProxier serves all HTTP endpoints by proxying
```
Comment (alesstimec) [inline internal/jujuapi/streamproxy.go]: due to implementation of the checkPermission, this only serves the /log endpoint atm

File: internal/jujuapi/streamproxy.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+		return
+	}
+
+	model, err := s.jimm.GetModel(ctx, uuid)
+	if err != nil {
+		writeError(err.Error(), errors.CodeModelNotFound)
+		return
+	}
+
+	if ok, err := checkPermission(ctx, finalPath, user, model.ResourceTag()); err != nil {
```
Comment (alesstimec) [inline internal/jujuapi/streamproxy.go]: we could do this permission check before fetching the controller on L74 since we have the model UUID

File: internal/jujuclient/dial.go [hunk 19]
```diff
@@ -331,7 +340,23 @@ func (c *Connection) Context() context.Context {
...
+	modelTag, ok := c.ModelTag()
+	if !ok {
+		return nil, errors.E(op, "no model found")
+	}
+
+	user, pass, err := c.dialer.ControllerCredentialsStore.GetControllerCredentials(c.ctx, c.ctl.Name)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	requestHeader := jujuhttp.BasicAuthHeader(names.NewUserTag(user).String(), pass)
```
Comment (alesstimec) [inline internal/jujuclient/dial.go:353]: please investigate if we can use the JWT to authenticate with the juju controller instead of basic auth..

File: internal/rpc/streamproxy.go [hunk 20]
```diff
@@ -0,0 +1,47 @@
+// Copyright 2024 Canonical.
+package rpc
```
Comment (alesstimec) [inline internal/rpc/streamproxy.go:3]: same here, we need a new line between copyright and package

File: internal/jujuapi/streamproxy_test.go (hunk 11 above)
Comment (ale8k) [reply internal/jujuapi/streamproxy_test.go:29]: Sure sure

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (ale8k) [reply internal/jujuapi/api_test.go:35]: Are you sure?

**Review by ale8k — DISMISSED** [review]

File: internal/jujuapi/api_test.go (hunk 9 above)
Comment (kian99) [reply internal/jujuapi/api_test.go:35]: That's what I recall happening, but this was like 2 months ago. I'll try it again now and let you know.

File: internal/auth/oauth2.go (hunk 13 above)
Comment (kian99) [reply internal/auth/oauth2.go:340]: Yes I'll add a godoc

File: internal/cmdtest/jimmsuite.go (hunk 14 above)
Comment (kian99) [reply internal/cmdtest/jimmsuite.go:161]: I agree, and that's my bad. I think I'd initially kept these changes separated into a specific commit but there are now a lot of commits.
I made the change because the CLI tests were broken when I upgraded the Juju dependency.

File: internal/jimm/model.go (hunk 12 above)
Comment (kian99) [reply internal/jimm/model.go:690]: Yes

File: internal/jujuclient/dial.go (hunk 19 above)
Comment (kian99) [reply internal/jujuclient/dial.go:353]: Looking at this with Simone, it's not implemented on Juju's end. So we'll need to keep this as is for now until we have Juju controller's that support JWT for HTTP auth.

**Review by SimoneDutto — APPROVED** [review]

**Review by alesstimec — APPROVED** [review]
